effects/butterfly-gears/butterfly-gears.c

```diff
@@ -0,0 +1,167 @@
+    CopWait(cp, y, COPWAIT_X);
+    CopMove16(cp, color[ 3 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 5 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 7 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 9 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[11 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[13 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[15 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[18 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[20 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[23 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[26 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[28 & COLREG_MASK], 0xf00);
+    CopMoveWord(cp, 0x1fe, 0);
+    y += 2;
+    CopWait(cp, y, COPWAIT_X);
+    CopMove16(cp, color[ 2 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 4 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 6 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[ 8 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[10 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[12 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[14 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[16 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[19 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[22 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[24 & COLREG_MASK], 0xf00);
+    CopMove16(cp, color[27 & COLREG_MASK], 0xf00);
+    CopMoveWord(cp, 0x1fe, 0);
+    y += 2;
+  }
+}
+
+static void MakeBallCopperList(BallCopListT *ballCp) {
+  CopListT *cp = NewCopList(INSTRUCTIONS_PER_BALL * 2 + 100);
+  ballCp->cp = cp;
+  CopInit(cp);
+  CopSetupDisplayWindow(cp, MODE_LORES, X(0), Y0, 320, 280);
+  CopMove16(cp, dmacon, DMAF_SETCLR | DMAF_RASTER);
+  CopLoadPal(cp, &testscreen_pal, 0);
+  CopSetupMode(cp, MODE_LORES, testscreen.depth);
+  CopSetupBitplanes(cp, NULL, &testscreen, testscreen.depth);
+  CopSetupBitplaneFetch(cp, MODE_LORES, X(0), testscreen.width);
+
+  ballCp->upperBallCopper = cp->curr;
+  InitCopperListBall(cp, Y0 + 7);
+
+  ballCp->lowerBallCopper = cp->curr;
+  InitCopperListBall(cp, Y0 + 127);
+
+  CopEnd(cp);
+}
+
+static void Init(void) {
+  MakeBallCopperList(&ballCopList1);
+  CopListActivate(ballCopList1.cp);
+  ball1.texture = texture_butterfly;
+  ball1.angleDelta = 27;
+  ball1.uvDelta = 1 << 25;
+  ball2.texture = texture_butterfly2;
+  ball2.angleDelta = -11;
+  ball2.uvDelta = -3 << (6+16);
+}
+
+static void Kill(void) {
+  DeleteCopList(ballCopList1.cp);
+}
+
+extern void PlotTextureAsm(char *copperDst asm("a0"),
+                           char *texture   asm("a1"),
+                           int  uvPosition asm("d5"),
+                           int  uvDeltaRow asm("d6"),
+                           int  uvDeltaCol asm("d1"));
+
+// Pack texture coordinates into a longword. Format:
+// Bits = UUUUUUUu uuuuuuuu VVVVVVVv vvvvvvvv
+// i.e. 7 integer bits and 9 fractional bits each
+static uint32_t uv(short u, short v) {
+  int combined;
+  combined = (v & 0xffff) | ((u & 0xffff) << 16);
```

Comment:
If you need doing some assembly tricks in C, please consider learning _gcc assembly inlines_. Here I hope optimizer kicked in and generated `swap` instruction. Do you check generated code with `objdump -d`? It's a very useful technique of hand optimizing code without touching assembler source code.

Reply:
The compiler does generate a `swap`, luckily; took a stab at explicit `asm` templating, but need more research…